# HG changeset patch
# User Michael Goffioul <michael.goffioul@gmail.com>
# Date 1429483268 14400
#      Sun Apr 19 18:41:08 2015 -0400
# Node ID a8769ccb2c4e39b46fa09568cb2ef71578df4939
# Parent  d58ba8b9f709e72dce9345f9a875671270a43fdc
Fix nested classdef package lookup (bug #44856)

* load-path.cc (load_path::dir_info::is_package): Split name and call
recursively to handle nested packages correctly.

diff --git a/libinterp/corefcn/load-path.cc b/libinterp/corefcn/load-path.cc
--- a/libinterp/corefcn/load-path.cc
+++ b/libinterp/corefcn/load-path.cc
@@ -107,17 +107,32 @@ load_path::dir_info::update (void)
       std::string msg = fs.error ();
       warning ("load_path: %s: %s", dir_name.c_str (), msg.c_str ());
     }
 }
 
 bool
 load_path::dir_info::is_package (const std::string& name) const
 {
-  return package_dir_map.find (name) != package_dir_map.end ();
+  size_t pos = name.find ('.');
+
+  if (pos == std::string::npos)
+    return package_dir_map.find (name) != package_dir_map.end ();
+  else
+    {
+      std::string name_head = name.substr (0, pos);
+      std::string name_tail = name.substr (pos + 1);
+
+      const_package_dir_map_iterator it = package_dir_map.find (name_head);
+
+      if (it != package_dir_map.end ())
+        return it->second.is_package (name_tail);
+      else
+        return false;
+    }
 }
 
 void
 load_path::dir_info::initialize (void)
 {
   is_relative = ! octave_env::absolute_pathname (dir_name);
 
   dir_time_last_checked = octave_time (static_cast<time_t> (0));
